common/lib/src/realm_types.dart

```diff
@@ -104,35 +103,87 @@ class RealmStateError extends StateError implements RealmError {
 class Decimal128 {} // TODO Support decimal128 datatype https://github.com/realm/realm-dart/issues/725
 
 /// @nodoc
-class RealmObjectBaseMarker {}
+abstract class RealmObjectMarker {}
 
-/// @nodoc
-class RealmObjectMarker extends RealmObjectBaseMarker {}
-
-/// @nodoc
-class EmbeddedObjectMarker extends RealmObjectBaseMarker {}
-
-// Union type
-/// @nodoc
-class RealmAny {
-  final dynamic value;
+/// A type that can represent any valid realm data type, except collections and embedded objects.
+///
+/// You can use [RealmValue] to declare fields on realm models, in which case it must be non-nullable,
+/// but it can wrap a null-value. List of [RealmValue] (`List<RealmValue>`) are also legal.
+///
+/// [RealmValue] fields can be [Indexed]
+///
+/// ```dart
+/// @RealmModel()
+/// class _AnythingGoes {
+///   @Indexed()
+///   late RealmValue any;
+///   late List<RealmValue> manyAny;
+/// }
+///
+/// void main() {
+///   final realm = Realm(Configuration.local([AnythingGoes.schema]));
+///   realm.write(() {
+///     final something = realm.add(AnythingGoes(any: RealmValue.string('text')));
+///     something.manyAny.addAll([
+///       null,
+///       true,
+///       'text',
+///       42,
+///       3.14,
+///     ].map(RealmValue.from));
+///   });
+/// }
+/// ```
+class RealmValue {
+  final Object? value;
+  Type get type => value.runtimeType;
```

Comment:
What if `value` is `null`?

Reply:
`runtimeType` is a method on `Object?`, it is safe to call on `null`
```dart
void main() {
  print(null.runtimeType);
}
```
prints: `Null` which is the runtime type of a `null` expression